python/ray/serve/batching.py

```diff
@@ -183,34 +191,51 @@ async def _consume_func_generator(
                     elif result in [StopIteration, StopAsyncIteration]:
                         # User's code returned sentinel. No values left
                         # for caller. Terminate iteration for caller.
-                        future.set_exception(StopAsyncIteration)
+                        _set_exception_if_not_done(future, StopAsyncIteration)
                         next_futures.append(FINISHED_TOKEN)
```

Comment:
Why do we even need FINISHED_TOKEN?

Reply:
We need some way to track which requests in the batch are finished. We use a `FINISHED_TOKEN` in the `futures` list as a placeholder for the completed requests until the entire batch is finished.

Reply:
Yeah, on the server-side we iterate until exhaustion. But on the client side, when requests in the batch finish, they terminate immediately. That way, clients don't have to wait for the whole batch to finish before their requests finish.